File: packages/ios-sdk/Sources/GateFlowKit/Misc/DeviceHelper.swift
import Foundation

/// Collects device information for enrichment.
struct DeviceInfo: Codable {
    let appId: String
    let appVersion: String
    let appBuild: String
    let osVersion: String
    let deviceModel: String
    let locale: String
    let timezone: String
    let timestamp: String
}

enum DeviceHelper {
    static func collect() -> DeviceInfo {
        let bundle = Bundle.main
        let appId = bundle.bundleIdentifier ?? "unknown"
        let appVersion = bundle.infoDictionary?["CFBundleShortVersionString"] as? String ?? "unknown"
        let appBuild = bundle.infoDictionary?["CFBundleVersion"] as? String ?? "unknown"
        let osVersion = ProcessInfo.processInfo.operatingSystemVersionString
        let deviceModel = deviceModelCode()
        let locale = Locale.current.identifier
        let timezone = TimeZone.current.identifier
        let timestamp = ISO8601DateFormatter().string(from: .now)

        return DeviceInfo(
            appId: appId,
            appVersion: appVersion,
            appBuild: appBuild,
            osVersion: osVersion,
            deviceModel: deviceModel,
            locale: locale,
            timezone: timezone,
            timestamp: timestamp
        )
    }

    private static func deviceModelCode() -> String {
        var size = 0
        sysctlbyname("hw.machine", nil, &size, nil, 0)
        var machine = [CChar](repeating: 0, count: size)
        sysctlbyname("hw.machine", &machine, &size, nil, 0)
        return String(cString: machine)
    }
}

File: packages/android-sdk/gateflow/src/main/java/com/gateflow/sdk/Misc/DeviceHelper.kt
package com.gateflow.sdk.misc

import android.content.Context
import android.os.Build
import java.text.SimpleDateFormat
import java.util.*

/** Collects device information for enrichment. */
data class DeviceInfo(
    val appId: String,
    val appVersion: String,
    val appBuild: String,
    val osVersion: String,
    val deviceModel: String,
    val locale: String,
    val timezone: String,
    val timestamp: String,
)

object DeviceHelper {
    fun collect(context: Context): DeviceInfo {
        val pkg = context.packageManager.getPackageInfo(context.packageName, 0)
        val appId = context.packageName
        val appVersion = pkg.versionName ?: "unknown"
        val appBuild = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            pkg.longVersionCode.toString()
        } else {
            pkg.versionCode.toString()
        }
        val osVersion = "Android ${Build.VERSION.RELEASE} (API ${Build.VERSION.SDK_INT})"
        val deviceModel = "${Build.MANUFACTURER} ${Build.MODEL}"
        val locale = Locale.getDefault().toString()
        val timezone = TimeZone.getDefault().id
        val timestamp = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US).format(Date())

        return DeviceInfo(
            appId = appId,
            appVersion = appVersion,
            appBuild = appBuild,
            osVersion = osVersion,
            deviceModel = deviceModel,
            locale = locale,
            timezone = timezone,
            timestamp = timestamp,
        )
    }
}
